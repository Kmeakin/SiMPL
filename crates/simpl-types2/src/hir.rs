pub use simpl_syntax2::ast::{Ident, Lit};
use simpl_syntax2::{ast, parse, ParseError};

pub type ExprId = u32;

#[derive(Debug, Copy, Clone, Default)]
struct IdGen {
    counter: u32,
}

impl IdGen {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn next(&mut self) -> u32 {
        let x = self.counter;
        self.counter += 1;
        x
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        id: ExprId,
        val: Lit,
    },
    Var {
        id: ExprId,
        name: String,
    },
    If {
        id: ExprId,
        test: Box<Expr>,
        then_branch: Box<Expr>,
        else_branch: Box<Expr>,
    },
    Let {
        id: ExprId,
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Letrec {
        id: ExprId,
        bindings: Vec<(Ident, Expr)>,
        body: Box<Expr>,
    },
    Lambda {
        id: ExprId,
        params: Vec<Ident>,
        body: Box<Expr>,
    },
    App {
        id: ExprId,
        func: Box<Expr>,
        arg: Box<Expr>,
    },
}

impl Expr {
    pub fn from_ast(ast: ast::Expr) -> Expr {
        let mut gen = IdGen::new();

        Expr::from_ast_inner(ast, &mut gen)
    }

    pub fn from_str(src: &str) -> Result<Expr, ParseError> {
        let ast = parse(src)?;
        Ok(Expr::from_ast(ast))
    }

    fn from_ast_inner(ast: ast::Expr, gen: &mut IdGen) -> Expr {
        let id = gen.next();

        match ast {
            ast::Expr::Lit { val } => Expr::Lit { id, val },
            ast::Expr::Var { name } => Expr::Var { id, name },
            ast::Expr::If {
                box test,
                box then_branch,
                box else_branch,
            } => Expr::If {
                id,
                test: box Expr::from_ast_inner(test, gen),
                then_branch: box Expr::from_ast_inner(then_branch, gen),
                else_branch: box Expr::from_ast_inner(else_branch, gen),
            },
            ast::Expr::Let { bindings, box body } => Expr::Let {
                id,
                bindings: bindings
                    .into_iter()
                    .map(|(var, val)| (var, Expr::from_ast_inner(val, gen)))
                    .collect(),

                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::Letrec { bindings, box body } => Expr::Letrec {
                id,
                bindings: bindings
                    .into_iter()
                    .map(|(var, val)| (var, Expr::from_ast_inner(val, gen)))
                    .collect(),
                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::Lambda { params, box body } => Expr::Lambda {
                id,
                params,
                body: box Expr::from_ast_inner(body, gen),
            },
            ast::Expr::App { box func, box arg } => Expr::App {
                id,
                func: box Expr::from_ast_inner(func, gen),
                arg: box Expr::from_ast_inner(arg, gen),
            },
        }
    }
}
