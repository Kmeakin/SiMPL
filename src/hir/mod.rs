pub use crate::types::ty::Type;
use crate::{syntax::ast, types::ty::TypeVarGen};
use derive_more::Display;
pub use simple_symbol::Symbol;
use std::str::FromStr;

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: Symbol,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then: Box<Self>,
        els: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    Lambda {
        ty: Type,
        param: Param,
        body: Box<Self>,
    },
    App {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Symbol,
    pub ann: Option<Type>,
    pub val: Box<Expr>,
}

#[derive(Debug, Copy, Clone, PartialEq, Display)]
pub enum Lit {
    #[display(fmt = "{}", _0)]
    Bool(bool),
    #[display(fmt = "{}", _0)]
    Int(i64),
    #[display(fmt = "{}", _0)]
    Float(f64),
}

impl From<bool> for Lit {
    fn from(other: bool) -> Self {
        Self::Bool(other)
    }
}

impl From<i64> for Lit {
    fn from(other: i64) -> Self {
        Self::Int(other)
    }
}

impl From<f64> for Lit {
    fn from(other: f64) -> Self {
        Self::Float(other)
    }
}

impl From<ast::Lit> for Lit {
    fn from(other: ast::Lit) -> Self {
        match other {
            ast::Lit::Int(x) => Self::Int(x),
            ast::Lit::Float(x) => Self::Float(x),
            ast::Lit::Bool(x) => Self::Bool(x),
        }
    }
}

impl FromStr for Expr {
    type Err = String;
    fn from_str(src: &str) -> Result<Self, Self::Err> {
        // TODO: return a trait object instead of unwrapping
        let ast = crate::syntax::parse(src).unwrap();
        Ok(Self::from_ast(ast))
    }
}

impl Expr {
    /// `ast::Expr` -> `hir::Expr`
    /// Expands nested let/lambda, and attatches fresh type variables to every
    /// expression/binder
    pub fn from_ast(ast: ast::Expr) -> Self {
        let mut gen = TypeVarGen::new();
        Self::from_ast_inner(ast, &mut gen)
    }

    fn from_ast_inner(ast: ast::Expr, gen: &mut TypeVarGen) -> Self {
        match ast {
            ast::Expr::Lit { val } => Self::Lit {
                val: val.into(),
                ty: gen.next(),
            },
            ast::Expr::Var { name } => Self::Var {
                name,
                ty: gen.next(),
            },
            ast::Expr::Binop { .. } => todo!(),
            ast::Expr::If { test, then, els } => Self::If {
                ty: gen.next(),
                test: box Self::from_ast_inner(*test, gen),
                then: box Self::from_ast_inner(*then, gen),
                els: box Self::from_ast_inner(*els, gen),
            },
            ast::Expr::Let { bindings, body } => {
                let (binding, body) = expand_let(&bindings, *body);
                Self::Let {
                    ty: gen.next(),
                    binding: LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::Letrec { bindings, body } => Self::Letrec {
                ty: gen.next(),
                bindings: bindings
                    .into_iter()
                    .map(|binding| LetBinding {
                        ty: gen.next(),
                        name: binding.name,
                        ann: binding.ann,
                        val: box Self::from_ast_inner(*binding.val, gen),
                    })
                    .collect(),
                body: box Self::from_ast_inner(*body, gen),
            },
            ast::Expr::Lambda { params, body } => {
                let (param, body) = expand_lambda(&params, *body);
                Self::Lambda {
                    ty: gen.next(),
                    param: Param {
                        name: param.name,
                        ann: param.ann,
                        ty: gen.next(),
                    },
                    body: box Self::from_ast_inner(body, gen),
                }
            }
            ast::Expr::App { func, arg } => Self::App {
                ty: gen.next(),
                func: box Self::from_ast_inner(*func, gen),
                arg: box Self::from_ast_inner(*arg, gen),
            },
        }
    }

    pub fn ty(&self) -> Type {
        match self {
            Self::Lit { ty, .. }
            | Self::Var { ty, .. }
            | Self::If { ty, .. }
            | Self::Let { ty, .. }
            | Self::Letrec { ty, .. }
            | Self::Lambda { ty, .. }
            | Self::App { ty, .. } => ty.clone(),
        }
    }
}

fn expand_lambda(params: &[ast::Param], body: ast::Expr) -> (ast::Param, ast::Expr) {
    assert!(!params.is_empty());
    if params.len() == 1 {
        (params[0].clone(), body)
    } else {
        let param = &params[0];
        let (rest_param, rest_body) = expand_lambda(&params[1..], body);
        (
            param.clone(),
            ast::Expr::Lambda {
                params: vec![rest_param],
                body: box rest_body,
            },
        )
    }
}

fn expand_let(bindings: &[ast::LetBinding], body: ast::Expr) -> (ast::LetBinding, ast::Expr) {
    assert!(!bindings.is_empty());
    if bindings.len() == 1 {
        (bindings[0].clone(), body)
    } else {
        let binding = &bindings[0];
        let (rest_bindings, rest_body) = expand_let(&bindings[1..], body);
        (
            binding.clone(),
            ast::Expr::Let {
                bindings: vec![rest_bindings],
                body: box rest_body,
            },
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::syntax::ast::{Expr, Lit};
    use simple_symbol::intern;

    #[test]
    fn test_expand_lambda() {
        let params = vec![
            ast::Param {
                name: intern("x"),
                ann: None,
            },
            ast::Param {
                name: intern("y"),
                ann: None,
            },
        ];
        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_lambda(&params, body),
            (
                ast::Param {
                    name: intern("x"),
                    ann: None
                },
                Expr::Lambda {
                    params: vec![ast::Param {
                        name: intern("y"),
                        ann: None
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }

    #[test]
    fn test_expand_let() {
        let bindings = vec![
            ast::LetBinding {
                name: intern("x"),
                val: box Expr::Lit { val: Lit::Int(1) },
                ann: None,
            },
            ast::LetBinding {
                name: intern("y"),
                val: box Expr::Lit { val: Lit::Int(2) },
                ann: None,
            },
        ];

        let body = Expr::Lit { val: Lit::Int(0) };

        assert_eq!(
            expand_let(&bindings, body),
            (
                ast::LetBinding {
                    name: intern("x"),
                    val: box Expr::Lit { val: Lit::Int(1) },
                    ann: None,
                },
                Expr::Let {
                    bindings: vec![ast::LetBinding {
                        name: intern("y"),
                        val: box Expr::Lit { val: Lit::Int(2) },
                        ann: None,
                    }],
                    body: box Expr::Lit { val: Lit::Int(0) }
                }
            ),
        )
    }
}
