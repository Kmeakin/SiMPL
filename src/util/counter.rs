use std::marker::PhantomData;

#[derive(Debug, Copy, Clone, Default)]
pub struct Counter<T>
where
    T: FromId,
{
    counter: u32,
    _phantom: PhantomData<T>,
}

impl<T: FromId> Counter<T> {
    pub fn new() -> Self {
        Self {
            counter: 0,
            _phantom: PhantomData,
        }
    }

    fn next_id(&mut self) -> u32 {
        let x = self.counter;
        self.counter += 1;
        x
    }

    pub fn current_id(&self) -> u32 {
        self.counter
    }

    pub fn current(&self) -> T {
        T::from_id(self.current_id())
    }

    pub fn next(&mut self) -> T {
        T::from_id(self.next_id())
    }
}

pub trait FromId {
    fn from_id(id: u32) -> Self;
}

impl FromId for u32 {
    fn from_id(id: u32) -> Self {
        id
    }
}
