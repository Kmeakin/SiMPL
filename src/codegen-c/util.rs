use crate::hir::{Expr, Type};
use maplit::{hashmap as hmap, hashset as hset};
use simple_symbol::Symbol;
use std::collections::{HashMap, HashSet};

pub type FreeVars = HashMap<Symbol, Type>;

pub fn free_vars(expr: &Expr) -> FreeVars {
    match expr {
        Expr::Lit { .. } => hmap![],
        Expr::Var { name, ty } => hmap![*name => ty.clone()],
        Expr::If {
            test, then, els, ..
        } => hashmap_union(
            hashmap_union(free_vars(test), free_vars(then)),
            free_vars(els),
        ),
        Expr::Let { binding, body, .. } => hashmap_diff(
            hashmap_union(free_vars(&*binding.val), free_vars(body)),
            hmap![binding.name => binding.ty.clone()],
        ),

        Expr::Letrec { bindings, body, .. } => hashmap_diff(
            bindings.iter().fold(free_vars(body), |acc, b| {
                hashmap_union(acc, free_vars(&*b.val))
            }),
            bindings.iter().map(|b| (b.name, b.ty.clone())).collect(),
        ),
        Expr::Lambda { param, body, .. } => {
            hashmap_diff(free_vars(body), hmap![param.name => param.ty.clone()])
        }

        Expr::App { func, arg, .. } => hashmap_union(free_vars(func), free_vars(arg)),
    }
}

fn hashmap_union<K, V>(hm1: HashMap<K, V>, hm2: HashMap<K, V>) -> HashMap<K, V>
where
    K: std::hash::Hash + Eq,
    V: std::hash::Hash + Eq,
{
    let mut ret = HashMap::new();
    ret.extend(hm1);
    ret.extend(hm2);
    ret
}

fn hashmap_diff<K, V>(hm1: HashMap<K, V>, hm2: HashMap<K, V>) -> HashMap<K, V>
where
    K: std::hash::Hash + Eq + Clone + std::fmt::Debug,
    V: std::hash::Hash + Eq + Clone + std::fmt::Debug,
{
    hm1.into_iter()
        .filter(|(k, _)| !hm2.contains_key(&k))
        .map(|(k, v)| (k.clone(), v.clone()))
        .collect::<HashMap<K, V>>()
}

#[cfg(test)]
mod test {
    use super::*;
    use maplit::hashset as hset;
    use simple_symbol::intern;
    use std::str::FromStr;

    #[track_caller]
    fn test_free_vars(src: &str, expected: HashSet<Symbol>) {
        let ast = Expr::from_str(src).unwrap();
        let free_vars = free_vars(&ast);
        let free_vars: HashSet<Symbol> = free_vars.iter().map(|(name, ty)| *name).collect();
        assert_eq!(free_vars, expected);
    }

    #[test]
    fn free_vars_lit() {
        test_free_vars("1", hset![]);
    }

    #[test]
    fn free_vars_if() {
        test_free_vars(
            "if abc then def else ghi",
            hset![intern("abc"), intern("def"), intern("ghi")],
        );
    }

    #[test]
    fn free_vars_let() {
        test_free_vars("let x = 5 in x y", hset![intern("y")]);
    }

    #[test]
    fn free_vars_letrec() {
        // TODO
    }

    #[test]
    fn free_vars_lambda() {
        test_free_vars(r"\x -> x", hset![]);
        test_free_vars(r"\x -> y", hset![intern("y")]);
    }

    #[test]
    fn free_vars_app() {
        test_free_vars("f x", hset![intern("f"), intern("x")]);
    }
}
