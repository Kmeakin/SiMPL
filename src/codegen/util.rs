use crate::hir::Expr;
use maplit::hashset as hset;
use simple_symbol::Symbol;
use std::collections::HashSet;

pub fn free_vars(expr: &Expr) -> HashSet<Symbol> {
    match expr {
        Expr::Lit { .. } => HashSet::new(),
        Expr::Var { name, .. } => hset![*name],
        Expr::If {
            test,
            then_branch,
            else_branch,
            ..
        } => &(&free_vars(test) | &free_vars(then_branch)) | &free_vars(else_branch),
        Expr::Let { binding, body, .. } => {
            &free_vars(&*binding.val) | &(&free_vars(body) - &hset!(binding.name))
        }
        Expr::Letrec { bindings, body, .. } => {
            &bindings
                .iter()
                .fold(free_vars(body), |acc, b| &acc | &free_vars(&*b.val))
                - &bindings.iter().map(|b| b.name).collect::<HashSet<_>>()
        }
        Expr::Lambda { param, body, .. } => &free_vars(body) - &hset!(param.name),
        Expr::App { func, arg, .. } => &free_vars(func) | &free_vars(arg),
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use maplit::hashset as hset;
    use simple_symbol::intern;
    use std::str::FromStr;

    #[track_caller]
    fn test_free_vars(src: &str, expected: HashSet<Symbol>) {
        let ast = Expr::from_str(src).unwrap();
        let free = free_vars(&ast);
        assert_eq!(free, expected);
    }

    #[test]
    fn free_vars_lit() {
        test_free_vars("1", hset![]);
    }

    #[test]
    fn free_vars_if() {
        test_free_vars(
            "if abc then def else ghi",
            hset![intern("abc"), intern("def"), intern("ghi")],
        );
    }

    #[test]
    fn free_vars_let() {
        test_free_vars("let x = 5 in x y", hset![intern("y")]);
    }

    #[test]
    fn free_vars_letrec() {
        // TODO
    }

    #[test]
    fn free_vars_lambda() {
        test_free_vars(r"\x -> x", hset![]);
        test_free_vars(r"\x -> y", hset![intern("y")]);
    }

    #[test]
    fn free_vars_app() {
        test_free_vars("f x", hset![intern("f"), intern("x")]);
    }
}
