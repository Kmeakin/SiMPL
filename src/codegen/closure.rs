//! Closure-conversion
//! Requires the `ANF` pass to be run first

use super::util::free_vars;
use crate::hir::{Expr, Lit, Param, Type};
use simple_symbol::Symbol;
use std::collections::{HashMap, HashSet};

mod pp;

type Env = HashMap<Symbol, CExpr>;

#[derive(Debug, Clone, PartialEq)]
pub struct LetBinding {
    pub ty: Type,
    pub name: Symbol,
    pub val: Box<CExpr>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum CExpr {
    Lit {
        ty: Type,
        val: Lit,
    },
    Var {
        ty: Type,
        name: Symbol,
    },
    If {
        ty: Type,
        test: Box<Self>,
        then_branch: Box<Self>,
        else_branch: Box<Self>,
    },
    Let {
        ty: Type,
        binding: LetBinding,
        body: Box<Self>,
    },
    Letrec {
        ty: Type,
        bindings: Vec<LetBinding>,
        body: Box<Self>,
    },
    MkClosure {
        ty: Type,
        param: Param,
        free_vars: HashSet<Symbol>,
        body: Box<Self>,
    },
    AppClosure {
        ty: Type,
        func: Box<Self>,
        arg: Box<Self>,
    },
    EnvRef {
        name: Symbol,
    },
}

fn substitute(cexpr: CExpr, subst: &HashMap<Symbol, CExpr>) -> CExpr {
    match cexpr {
        CExpr::Lit { .. } => cexpr,
        CExpr::Var { ty, name } => subst.get(&name).unwrap_or(&CExpr::Var { ty, name }).clone(),
        CExpr::If {
            ty,
            test,
            then_branch,
            else_branch,
        } => CExpr::If {
            ty: ty.clone(),
            test: box substitute(*test, subst),
            then_branch: box substitute(*then_branch, subst),
            else_branch: box substitute(*else_branch, subst),
        },
        CExpr::Let { ty, binding, body } => CExpr::Let {
            ty: ty.clone(),
            binding: LetBinding {
                val: box substitute(*binding.val, subst),
                ..binding
            },
            body: box substitute(*body, subst),
        },
        CExpr::Letrec { .. } => todo!(),
        CExpr::MkClosure {
            ty,
            param,
            free_vars,
            body,
        } => CExpr::MkClosure {
            ty,
            param,
            free_vars,
            body: box substitute(*body, subst),
        },
        CExpr::AppClosure { ty, func, arg } => CExpr::AppClosure {
            ty,
            func: box substitute(*func, subst),
            arg: box substitute(*arg, subst),
        },
        CExpr::EnvRef { name } => CExpr::EnvRef { name },
    }
}

fn closure_convert(expr: &Expr) -> CExpr {
    match expr {
        Expr::Lit { ty, val } => CExpr::Lit {
            ty: ty.clone(),
            val: *val,
        },
        Expr::Var { ty, name } => CExpr::Var {
            ty: ty.clone(),
            name: *name,
        },
        Expr::If {
            ty,
            test,
            then_branch,
            else_branch,
        } => CExpr::If {
            ty: ty.clone(),
            test: box closure_convert(&**test),
            then_branch: box closure_convert(&**then_branch),
            else_branch: box closure_convert(&**else_branch),
        },
        Expr::Let { ty, binding, body } => CExpr::Let {
            ty: ty.clone(),
            binding: LetBinding {
                ty: binding.ty.clone(),
                name: binding.name,
                val: box closure_convert(&*binding.val),
            },
            body: box closure_convert(&*body),
        },
        Expr::Letrec { .. } => todo!(),
        Expr::Lambda {
            ty,
            param,
            ref body,
        } => {
            let fv = free_vars(&expr.clone());
            CExpr::MkClosure {
                ty: ty.clone(),
                param: param.clone(),
                free_vars: fv.clone(),
                body: box substitute(
                    closure_convert(&**body),
                    dbg!(&fv
                        .iter()
                        .map(|&name| (name, CExpr::EnvRef { name }))
                        .collect()),
                ),
            }
        }
        Expr::App { ty, func, arg } => CExpr::AppClosure {
            ty: ty.clone(),
            func: box closure_convert(&**func),
            arg: box closure_convert(&**arg),
        },
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::{
        codegen::anf::normalize_expr,
        hir::Expr,
        types::{self, parse_and_type},
    };
    use insta::assert_snapshot;
    use std::str::FromStr;

    #[track_caller]
    fn test_closure_convert(src: &str) {
        let expr = Expr::from_str(src).unwrap();
        let anf = normalize_expr(expr);
        let clo_conv = closure_convert(&anf);
        dbg!(&clo_conv);
        assert_snapshot!(clo_conv.pretty());
    }

    #[test]
    fn test0() {
        let src = r"
let const = \x -> \ignored -> x
in const 5
";
        test_closure_convert(src);
    }

    #[test]
    fn test1() {
        let src = r"
let x = 5,
    f = \ignored -> x
in f 100
";
        test_closure_convert(src);
    }

    #[test]
    fn test2() {
        let src = r"
let x = 5,
    y = 10,
    f = \ignored -> add y x
in f 100
";
        test_closure_convert(src);
    }
}
