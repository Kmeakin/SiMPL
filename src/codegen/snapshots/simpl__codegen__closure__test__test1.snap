---
source: src/codegen/closure.rs
expression: clo_conv.pretty()
---
let x = 5 in
    let f = \ignored -> (envRef x) in
        (f 100)
